crates/red_knot_python_semantic/src/types.rs

```diff
@@ -5650,31 +5725,65 @@ impl<'db> BoolError<'db> {
+                    .into_lookup_result()
+                    .ok()
+                    .and_then(|quals| quals.inner_type().parameter_span(context.db(), None))
+                {
+                    sub.annotate(
+                        Annotation::primary(parameter_span).message("Incorrect parameters"),
+                    );
+                    sub.annotate(Annotation::secondary(func_span).message("Method defined here"));
+                }
+                diag.sub(sub);
             }
             Self::IncorrectReturnType {
                 not_boolable_type,
                 return_type,
             } => {
-                builder.into_diagnostic(format_args!(
-                    "Boolean conversion is unsupported for type `{not_boolable}`; \
-                     the return type of its bool method (`{return_type}`) \
-                     isn't assignable to `bool",
+                let mut diag = builder.into_diagnostic(format_args!(
+                    "Boolean conversion is unsupported for type `{not_boolable}`",
                     not_boolable = not_boolable_type.display(context.db()),
-                    return_type = return_type.display(context.db())
                 ));
+                let mut sub = SubDiagnostic::new(
+                    Severity::Info,
+                    format_args!(
+                        "`{return_type}` is not assignable to `bool`",
+                        return_type = return_type.display(context.db()),
+                    ),
+                );
+                if let Some((func_span, return_type_span)) = not_boolable_type
+                    .member(context.db(), "__bool__")
+                    .into_lookup_result()
+                    .ok()
+                    .and_then(|quals| quals.inner_type().return_type_span(context.db()))
+                {
+                    sub.annotate(
+                        Annotation::primary(return_type_span).message("Incorrect return type"),
+                    );
+                    sub.annotate(Annotation::secondary(func_span).message("Method defined here"));
+                }
+                diag.sub(sub);
             }
             Self::NotCallable { not_boolable_type } => {
-                builder.into_diagnostic(format_args!(
-                    "Boolean conversion is unsupported for type `{}`; \
-                     its `__bool__` method isn't callable",
+                let mut diag = builder.into_diagnostic(format_args!(
+                    "Boolean conversion is unsupported for type `{}`",
                     not_boolable_type.display(context.db())
                 ));
+                let sub = SubDiagnostic::new(Severity::Info, "`__bool__` must be callable");
+                // TODO: It would be nice to create an annotation here for
+                // where `__bool__` is defined. At time of writing, I couldn't
+                // figure out a straight-forward way of doing this. ---AG
```

Comment:
Yeah, this has overlap with go-to-definition; it requires adding APIs to not just get a symbol, but get the location(s) of its definition(s), and we don't currently have those APIs. I don't think they are inherently difficult to add over what we have now (the information is all there in the use-def map, it's just a Simple Matter of Programming), but it's a non-trivial amount of Programming.

cc @MichaReiser

Reply:
I'm on the run so I haven't looked closely at the code but we now have `Type::definition`... in case you have a type pointing to `__bool__` somewhere.

Reply:
That's true, in the common case where `__bool__` is a method, the `FunctionType::definition` ought to point to where it was defined?

Reply:
Yeah this is all pretty much the same conclusion I arrived at. I had indeed found `Type::definition`.

I'm at least happy that I wasn't missing something dumb. :-)

(Totally agree about this being a "simple matter of programming.")